doc/man/tabs.1
.Dd December 2008
.Dt TABS 1
.Sh NAME
.Nm tabs
.Nd set terminal tabs
.Sh SYNOPSIS
.Nm
.Op "" Fl Ns Ar n Ns | Fl a| Fl a2| Fl c| Fl c2| Fl c3| Fl f| Fl p| Fl s| Fl u
.Op \+m Ns Op n
.Op Fl T Ar type
.Pp
.Nm
.Op Fl T Ar type Ns
.Op ""\+ Ns Op Ar n
.Ar n1 Ns Op ,n2,...
.Sh DESCRIPTION
The
.Nm
utility will display a series of characters that first clears
the hardware terminal tab settings and then initializes the tab stops at
the specified positions [XSI] [Option Start] and optionally adjusts the
margin.  [Option End]
.Pp
The phrase
.Qq tab-stop position Ar N
will be taken to mean that, from the start of a line of output, tabbing
to position
.Ar N
will cause the next character output to be in the
.Pq Ar N Ns +1 Ns
th column position on that line.  The maximum number of tab stops
allowed is terminal-dependent.
.Pp
It need not be possible to implement
.Nm
on certain terminals.  If the terminal type obtained from the
.Ev TERM
environment variable or
.Fl T
option represents such a terminal, an appropriate diagnostic message
will be written to standard error and tabs will exit with a status
greater than zero.
.Sh OPTIONS
The
.Nm
utility will conform to the Base Definitions volume of IEEE Std
1003.1-2001, Section 12.2, Utility Syntax Guidelines, [XSI] [Option
Start] except for various extensions: the options
.Fl a2 , c2
and
.Fl c3
are multi-character.  [Option End]
.Pp
The following options will be supported:
.Bl -tag -width flags
.It Fl Ar n
Specify repetitive tab stops separated by a uniform number of column
positions,
.Ar n ,
where
.Ar n
is a single-digit decimal number.  The default usage of
.Nm
with no arguments will be equivalent to
.Nm Fl Ar 8 .
When
.Fl Ar 0
is used, the tab stops will be cleared and no new ones set.
.It Fl a
[XSI] [Option Start] 1,10,16,36,72
.Pp
Assembler, applicable to some mainframes.  [Option End]
.It Fl a2
[XSI] [Option Start] 1,10,16,40,72
.Pp
Assembler, applicable to some mainframes.  [Option End]
.It Fl c
[XSI] [Option Start] 1,8,12,16,20,55
.Pp
COBOL, normal format.  [Option End]
.It Fl c2
[XSI] [Option Start] 1,6,10,14,49
.Pp
COBOL, compact format (columns 1 to 6 omitted).  [Option End]
.It Fl c3
[XSI] [Option Start] 1,6,10,14,18,22,26,30,34,38,42,46,50,54,58,62,67
.Pp
COBOL compact format (columns 1 to 6 omitted), with more tabs than
.Fl c2 .
[Option End]
.It Fl f
[XSI] [Option Start] 1,7,11,15,19,23
.Pp
FORTRAN [Option End]
.It Fl p
[XSI] [Option Start] 1,5,9,13,17,21,25,29,33,37,41,45,49,53,57,61
.Pp
PL/1 [Option End]
.It Fl s
[XSI] [Option Start] 1,10,55
.Pp
SNOBOL [Option End]
.It Fl u
[XSI] [Option Start] 1,12,20,44
.Pp
Assembler, applicable to some mainframes.  [Option End]
.It Fl T Ar type
Indicate the type of terminal.  If this option is not supplied and the
.Ev TERM
variable is unset or null, an unspecified default terminal type will be
used.  The setting of type will take precedence over the value in
.Ev TERM .
.El
.Sh OPERANDS
.Pp
The following operand will be supported:
.Bl -tag -width
.It Ar n1 Ns Op , Ns Ar n2 , Ns ...
A single command line argument that consists of tab-stop values
separated using either commas or <blank>s.  The application will ensure
that the tab-stop values are positive decimal integers in strictly
ascending order.  If any number (except the first one) is preceded by a
plus sign, it is taken as an increment to be added to the previous
value.  For example, the tab lists 1,10,20,30 and 1,10,+10,+10 are
considered to be identical.
.El
.Sh STDIN
Not used.
.Sh INPUT FILES
None.
.Sh ENVIRONMENT VARIABLES
The following environment variables will affect the execution of tabs:
.Bl -tag -width
.It Ev LANG
Provide a default value for the internationalization variables that are
unset or null.  (See the Base Definitions volume of IEEE Std
1003.1-2001, Section 8.2, Internationalization Variables for the
precedence of internationalization variables used to determine the
values of locale categories.)
.It Ev LC_ALL
If set to a non-empty string value, override the values of all the other
internationalization variables.
.It Ev LC_CTYPE
Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to
multi-byte characters in arguments).
.It Ev LC_MESSAGES
Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
.It Ev NLSPATH
[XSI] [Option Start] Determine the location of message catalogs for the
processing of
.Ev LC_MESSAGES .
[Option End]
.Pp
.It Ev TERM
Determine the terminal type.  If this variable is unset or null, and if
the
.Fl T
option is not specified, an unspecified default terminal type
will be used.
.El
.Sh ASYNCHRONOUS EVENTS
Default.
.Sh STDOUT
If standard output is a terminal, the appropriate sequence to clear and
set the tab stops may be written to standard output in an unspecified
format.  If standard output is not a terminal, undefined results occur.
.Sh STDERR
The standard error will be used only for diagnostic messages.
.Sh OUTPUT FILES
None.
.Sh EXTENDED DESCRIPTION
None.
.Sh EXIT STATUS
The following exit values will be returned:
.Bl -tag -width
.It 0
Successful completion.
.It >0
An error occurred.
.El
.Sh CONSEQUENCES OF ERRORS
Default.
.Pp
The following sections are informative.
.Sh APPLICATION USAGE
This utility makes use of the terminal's hardware tabs and the stty tabs
option.
.Pp
This utility is not recommended for application use.
.Pp
Some integrated display units might not have escape sequences to set tab
stops, but may be set by internal system calls.  On these terminals,
.Nm
works if standard output is directed to the terminal; if output is
directed to another file, however,
.Nm
fails.
.Sh EXAMPLES
None.
.Sh RATIONALE
Consideration was given to having the tput utility handle all of the
functions described in
.Nm .
However, the separate
.Nm
utility was retained because it seems more intuitive to use a command
named
.Nm
than tput with a new option.  The tput utility does not support setting
or clearing tabs, and no known historical version of
.Nm
supports the capability of setting arbitrary tab stops.
.Pp
The System V tabs interface is very complex; the version in this volume
of IEEE Std 1003.1-2001 has a reduced feature list, but many of the
features omitted were restored as XSI extensions even though the
supported languages and coding styles are primarily historical.
.Pp
There was considerable sentiment for specifying only a means of
resetting the tabs back to a known state-presumably the "standard" of
tabs every eight positions.  The following features were omitted:
.Pp
.Bl -bullet
.It
Setting tab stops via the first line in a file, using --
.Ar file .
Since even the SVID has no complete explanation of this feature, it is
doubtful that it is in widespread use.
.El
.Pp
In an early proposal, a
.Fl t
tablist option was added for consistency with
.Xr expand ;
this was later removed when inconsistencies with the historical list of
tabs were identified.
.Pp
Consideration was given to adding a
.Fl p
option that would output the current tab settings so that they could be
saved and then later restored.  This was not accepted because querying
the tab stops of the terminal is not a capability in historical terminfo
or termcap facilities and might not be supported on a wide range of
terminals.
.Sh FUTURE DIRECTIONS
None.
.Sh SEE ALSO
.Xr expand ,
.Xr stty ,
.Xr tput ,
.Xr unexpand


NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



tabs.c

/*
 * Copyright 2008-2009 Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <curses.h>
#include <term.h>
#include <libpu.h>

enum {
	tabs_outbuf_sz		= 4096,
	max_tab_stops		= 1024,
};

static char *opt_term;
static int n_tabs;
static int tab_stop[max_tab_stops];

static const int tabset_a[] = { 1,10,16,36,72 };
static const int tabset_a2[] = { 1,10,16,40,72 };
static const int tabset_c[] = { 1,8,12,16,20,55 };
static const int tabset_c2[] = { 1,6,10,14,49 };
static const int tabset_c3[] = { 1,6,10,14,18,22,26,30,34,38,42,46,50,54,58,62,67 };
static const int tabset_f[] = { 1,7,11,15,19,23 };
static const int tabset_p[] = { 1,5,9,13,17,21,25,29,33,37,41,45,49,53,57,61 };
static const int tabset_s[] = { 1,10,55 };
static const int tabset_u[] = { 1,12,20,44 };

static void usage(void)
{
	fprintf(stderr,
"Usage:\n"
"tabs [ -n| -a| -a2| -c| -c2| -c3| -f| -p| -s| -u][+m[n]] [-T type]\n"
"        or\n"
"tabs [-T type][ +[n]] n1[,n2,...]\n");
	exit(1);
}

static void tab_push(int stop)
{
	if (n_tabs == max_tab_stops) {
		fprintf(stderr, _("tabs: tab stop table limit reached\n"));
		exit(1);
	}

	tab_stop[n_tabs++] = stop;
}

#define tab_stock(arr) __tab_stock(arr, ARRAY_SIZE(arr))

static void __tab_stock(const int *stops, int n_stops)
{
	n_tabs = n_stops;
	memcpy(tab_stop, stops, n_stops * sizeof(int));
}

static void tab_repeating(int tab_n)
{
	int col;

	for (col = 1; col < columns; col++)
		if (col % tab_n == 0)
			tab_push(col);
}

static void cmdline(int argc, char **argv)
{
	int opt, i, col = 1, err = 0;
	bool changed = false;
	char tmparg[4096], *tok, *tmp_s;
	const int skip = 2;	/* skip -T */

	/* get terminal type immediately, so that calculations
	 * based on terminal size, needed during argument parsing,
	 * will succeed.
	 */
	for (i = 1; i < argc; i++)
		if (strncmp(argv[i], "-T", 2) == 0) {
			if (*(argv[i] + skip) == '\0') {
				i++;
				if (i < argc)
					opt_term = argv[i];
			} else {
				opt_term = argv[i] + skip;
			}
		}

	if (setupterm(opt_term, STDOUT_FILENO, &err) != OK) {
		fprintf(stderr, _("tabs: setupterm failed for term type %s\n"),
			(opt_term && *opt_term) ? opt_term : _("(null)"));
		exit(1);
	}

	tab_repeating(8);

	while ((opt = getopt(argc, argv, "0123456789a::c::fpsuT:")) != -1) {
		switch (opt) {
		case '0':
			n_tabs = 0;
			changed = true;
			break;

		case '1':
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
		case '8':
		case '9':
			tab_repeating(opt - '0');
			changed = true;
			break;

		case 'a':
			changed = true;
			if (!optarg)
				tab_stock(tabset_a);
			else if (!strcmp(optarg, "2"))
				tab_stock(tabset_a2);
			else
				usage();
			break;

		case 'c':
			changed = true;
			if (!optarg)
				tab_stock(tabset_c);
			else if (!strcmp(optarg, "2"))
				tab_stock(tabset_c2);
			else if (!strcmp(optarg, "3"))
				tab_stock(tabset_c3);
			else
				usage();
			break;

		case 'f': changed = true; tab_stock(tabset_f); break;
		case 'p': changed = true; tab_stock(tabset_p); break;
		case 's': changed = true; tab_stock(tabset_s); break;
		case 'u': changed = true; tab_stock(tabset_u); break;

		case 'T':
			break;

		default:
			usage();
			break;
		}
	}

	if (optind == argc)
		return;

	if (changed)
		usage();

	n_tabs = 0;

	while (optind < argc) {
		bool do_offset;
		int offset, last_col = -1;

		strcpy(tmparg, argv[optind++]);
		tmp_s = tmparg;
		while ((tok = strtok(tmp_s, ", \r\n\f\t")) != NULL) {
			tmp_s = NULL;
			do_offset = false;
			if (*tok == '+') {
				tok++;
				do_offset = true;
			}
			offset = atoi(tok);

			if (do_offset) {
				if (offset > columns ||
				    offset > (col + columns)) {
					fprintf(stderr, _("tabs: tabspec beyond max-columns specified\n"));
					exit(1);
				}

				col += offset;
			} else {
				if (offset < col)
					usage();

				col = offset;
			}

			if (col == last_col)
				usage();

			tab_push(col);

			last_col = col;
		}
	}
}

static char outbuf[tabs_outbuf_sz];
static int outbuf_avail = tabs_outbuf_sz - 1;

static bool push_str(const char *val)
{
	size_t val_len;

	if (!val)
		return true;

	val_len = strlen(val);
	if (outbuf_avail < val_len)
		return false;
	
	strcat(outbuf, val);
	outbuf_avail -= val_len;

	return true;
}

static void set_hw_tabs(void)
{
	int i, col = 0;
	char *tab_set;

	tab_set = xtigetstr("hts");
	if (!tab_set) {
		fprintf(stderr, _("tabs: terminal unable to set tabs\n"));
		exit(1);
	}

	if (!push_str(xtigetstr("tbc")))
		goto err_out;

	for (i = 0; i < n_tabs; i++) {
		while (col < tab_stop[i]) {
			if (!push_str(" "))
				goto err_out;
			col++;
		}

		if (!push_str(tab_set))
			goto err_out;
	}

	write(STDOUT_FILENO, outbuf, strlen(outbuf));

	return;

err_out:
	fprintf(stderr, _("tabs: buffer overflow\n"));
	exit(1);
}

int main (int argc, char *argv[])
{
	pu_init();
	opt_term = get_terminal();
	cmdline(argc, argv);
	set_hw_tabs();

	return 0;
}
